### ISSUE TITLE
Use node name, not uuid, whenever possible.

### ISSUE THREAD

**Issue by Kamilcuk** [hashicorp/nomad#17725]
### Proposal
Nomad node uuid are random and unpredictable and unmanageable and unreadable. Consider using node Name whenever possible.

In the UI when showing Job tasks, show node name.

In the cli `nomad job status` when showing allocations, show node name.

Allow `nomad node status -name somename` to get the node status of some node.

### Use-cases
The cases is that uuid are completely random and users nor devops team do not want to learn some random uuid. Instead, node name is the same as hostname, which is uniquely identifiable and fast understood and encodes the location of the server in a simple format.

### Attempted Solutions
The current solution is to wrap `nomad job status` and execute `nomad node status <uuid>` for each uuid and to hover with mouse over uuid in the ui.

**Comment by lgfa29** [comment]
Thanks for the suggestion @Kamilcuk. I agree that names are much more user-friendly, but there are a few issues we would need to handle:

1. Node names are not unique, so targeting a node by name may not generate the expected result.
2. Node names are not stable, meaning they can change throughout the node's lifecycle.
3. Because of 2, we only store the node ID in related objects (like evals and allocs), so in order to display the node name while viewing these entities we would need an extra query.

I have placed this in our board for further triaging and roadmapping. Thank you again for the suggestion!

**Comment by Kamilcuk** [comment]
Hi, thanks for response. My 2 cents:
1. Yes. Node names are for humans, humans will know when they have duplicates in a cluster and will resolve them using context. 
1.1. Also, similarly, job names are also not unique - they are unique within a namespace. For example,  `NOMAD_NAMESPACE='*' nomad job status <jobname>` fails when there are two jobs with the same name.
1.2. (For me, personally, node ids are confusing and unnecessary level of abstraction. I would enjoy node names beeing unique).
3. Yes. But humans will know when they change something and will synchronize together within a team.
4. I do not care about evaluations, but! It _has_ already been done for allocations and there is alloc.NodeName. Node name _is_ already displayed when you hover with mouse over the allocation on job allocation web page. I would want to switch them - show the node name and show node id when hovering with mouse.

Thanks!

**Comment by lgfa29** [comment]
> job names are also not unique - they are unique within a namespace

Yeah, the namespace is the key part (and one that we sometimes forget about 😅). A namespace is required to uniquely identify a job.

> For me, personally, node ids are confusing and unnecessary level of abstraction. I would enjoy node names beeing unique

That's fair. Unique IDs are important internally, but users may not need to care about them.

> Node name is already displayed when you hover with mouse over the allocation on job allocation web page.

Ha! I never noticed that we do store the node name in allocations 😅 
https://github.com/hashicorp/nomad/blob/d14362ec1903fa8d49db498c28cfdf4b5f073b22/nomad/structs/structs.go#L10348-L10349

One extra bit I would add is that querying node by name is not very efficient due to how our data storage works. You need to traverse all nodes to find the ones that match a give name. But that is probably manageable.

Accepting node names in CLI commands sound like a good start for this.

**Comment by Kamilcuk** [comment]
I started writing tampermonkey script to do that. How do I hook to emberjs to change the `a` content _after_ they are added? I am completely green in emberjs. The following works, but the 1 second delay is not nice.

```
(function() {
    'use strict';

    function log(a) {
        console.log("NOMADUISWAP: " + a);
    }

    // Function to find the closest span element with an aria-label attribute
    function findClosestSpan(element) {
        while (element.parentElement) {
            element = element.parentElement;
            if (element.tagName.toLowerCase() === 'span' && element.hasAttribute('aria-label')) {
                return element;
            }
        }
        return null;
    }

    // Function to modify the content of the a element
    function modifyContent() {
        var aElements = document.getElementsByTagName('a');
        log("found elements " + aElements.length);
        for (var i = 0; i < aElements.length; i++) {
            var aElement = aElements[i];
            var href = aElement.getAttribute('href');
            log("AAAAAAAAAAAAAA " + aElement);
            if (href && href.match(/\/ui\/clients\/.*/)) {
                var spanElement = findClosestSpan(aElement);
                if (spanElement) {
                    var ariaLabel = spanElement.getAttribute('aria-label');
                    aElement.textContent = ariaLabel;
                }
            }
        }
    }

    log("hello world");
    Ember.run.debounce('afterRender', modifyContent, 1000);
    //    Ember.run.scheduleOnce('afterRender', modifyContent);
    // modifyContent();
    // Run the function after the DOM is fully loaded
    // document.addEventListener('DOMContentLoaded', modifyContent);
})();
```

**Comment by Kamilcuk** [comment]
I was able to write the following tampermonkey script that shows node names instead of UUIDs in Nomad web ui:

```
(function() {
    'use strict';
    function log(a) {
        console.log("NOMADNODESWAP: " + a);
    }
    var observer = new MutationObserver(function(mutations) {
        mutations.forEach(function(mutation) {
            mutation.addedNodes.forEach(function(node) {
                // If element is a
                if (node.nodeType === Node.ELEMENT_NODE && node.nodeName.toLowerCase() === 'a') {
                    // If a has content that is uuid
                    if (node.textContent.match(/[0-9a-fA-F]{6,}/)) {
                        var href = node.getAttribute('href');
                        // If a href= points to ui/clients/uuid
                        if (href && href.match(/\/ui\/clients\/[0-9a-fA-F]{6,}/)) {
                            var parent = node.parentElement;
                            // If the parent is span with aria-label
                            if (parent && parent.tagName.toLowerCase() === 'span' && parent.hasAttribute('aria-label')) {
                                // Set the a content to aria-label of span label.
                                var ariaLabel = parent.getAttribute('aria-label');
                                node.textContent = ariaLabel;
                            }
                        }
                    }
                }
            });
        });
    });
    observer.observe(document.body, { childList: true, subtree: true });
    // Object.keys(Ember).forEach((prop)=> console.log(prop));
    // Ember.run.debounce('afterRender', modifyContent, 2000);
    // modifyContent();
    // Run the function after the DOM is fully loaded
    // document.addEventListener('DOMContentLoaded', modifyContent);
})();
```

Now all UUIDs are replaced with client names! Looks great.

**Comment by Kamilcuk** [comment]
Related https://github.com/hashicorp/nomad/issues/5177

**Comment by 116davinder** [comment]
@Kamilcuk, thanks for raising this, I have been annoyed with UUID since I started working with Nomad. I have to click 2 or 3 times on UUID to reach the real node name.

Thanks for the script, I will try this.
